translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,382 @@
+// Copyright 2020 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package opencensus
+
+import (
+	"strings"
+
+	occommon "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	ocresource "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	octrace "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/golang/protobuf/ptypes"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/internal"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	tracetranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+)
+
+func ocToInternal(td consumerdata.TraceData) data.TraceData {
+
+	if td.Node == nil && td.Resource == nil && len(td.Spans) == 0 {
+		return data.TraceData{}
+	}
+
+	resource := ocNodeResourceToInternal(td.Node, td.Resource)
+
+	resourceSpans := data.NewResourceSpans(resource, nil)
+	resourceSpanList := []*data.ResourceSpans{resourceSpans}
+
+	spanCount := len(td.Spans)
+	if spanCount == 0 {
+		return data.NewTraceData(resourceSpanList)
+	}
+
+	// Create slice that holds pointers and another slice that holds structs at once
+	// to avoid individual allocations of structs.
+	spans := data.NewSpanSlice(spanCount)
+	spanPtrs := make([]*data.Span, 0, spanCount)
```

Comment:
Probably data.NewSpanSlice should return the slice of pointers to match what the ResourceSpans uses? TODO or issue is fine.

Reply:
All new New*Slice functions return a slice of structs, not pointers. The idea is that we make one allocation. However we still allocate the slice of pointers separately because not every struct will end up in the same slice of pointers. If New*Slice returns a slice of pointers we will have to rebuild the slice in this function, which will likely be more confusing.